src/service_control_client_impl_quota_test.cc

```diff
@@ -0,0 +1,372 @@
+  ServiceControlClientOptions options(CheckAggregationOptions(0, 500, 1000),
+                                      QuotaAggregationOptions(10, 500),
+                                      ReportAggregationOptions(0, 500));
+
+  options.check_transport = mock_check_transport_.GetFunc();
+  options.quota_transport = mock_quota_transport_.GetFunc();
+  options.report_transport = mock_report_transport_.GetFunc();
+
+  int callbackCalledCount = 0;
+
+  std::unique_ptr<ServiceControlClient> client_ =
+      CreateServiceControlClient(kServiceName, kServiceConfigId, options);
+
+  Status done_status = Status::UNKNOWN;
+  AllocateQuotaResponse quota_response;
+
+  client_->Quota(quota_request1_, &quota_response,
+                 [&done_status](Status status) { done_status = status; },
+                 [&callbackCalledCount](const AllocateQuotaRequest& request,
+                                        AllocateQuotaResponse* response,
+                                        TransportDoneFunc on_done) {
+                   callbackCalledCount++;
+                   on_done(Status::OK);
+                 });
+
+  for (int i = 0; i < 10; i++) {
+    client_->Quota(quota_request1_, &quota_response,
+                   [&done_status](Status status) { done_status = status; },
+                   [&callbackCalledCount](const AllocateQuotaRequest& request,
+                                          AllocateQuotaResponse* response,
+                                          TransportDoneFunc on_done) {
+                     callbackCalledCount++;
+                     on_done(Status::OK);
+                   });
+  }
+
+  EXPECT_EQ(done_status, Status::OK);
+
+  EXPECT_EQ(callbackCalledCount, 1);
+
+  Statistics stat;
+  Status stat_status = client_->GetStatistics(&stat);
+
+  EXPECT_EQ(stat_status, Status::OK);
+  EXPECT_EQ(stat.total_called_quotas, 11);
+  EXPECT_EQ(stat.send_quotas_by_flush, 0);
+  EXPECT_EQ(stat.send_quotas_in_flight, 1);
+
+  EXPECT_EQ(stat.send_report_operations, 0);
+}
+
+TEST_F(ServiceControlClientImplQuotaTest, TestNonCachedQuotaInSync) {
+  EXPECT_CALL(mock_quota_transport_, Quota(_, _, _))
+      .WillRepeatedly(Invoke(&mock_quota_transport_,
+                             &MockQuotaTransport::AllocateQuotaUsingThread));
+
+  ServiceControlClientOptions options(CheckAggregationOptions(0, 500, 1000),
+                                      QuotaAggregationOptions(0, 500),
+                                      ReportAggregationOptions(0, 500));
+
+  options.check_transport = mock_check_transport_.GetFunc();
+  options.quota_transport = mock_quota_transport_.GetFunc();
+  options.report_transport = mock_report_transport_.GetFunc();
+
+  std::unique_ptr<ServiceControlClient> client_ =
+      CreateServiceControlClient(kServiceName, kServiceConfigId, options);
+
+  AllocateQuotaResponse quota_response;
+  Status done_status = client_->Quota(quota_request1_, &quota_response);
+  EXPECT_EQ(done_status, Status::OK);
```

Comment:
you have to use "future/promise" to wait for the done_status since it will be called from another thread.